### PR TITLE
feat: Unnormalize waiting requests + predictive load updates for Python router (mirroring Rust) + softmax sampling to reduce thrashing

### PR DESCRIPTION
**Pull request by PeaBrane**
#### Overview:

Fairly high-priority KV router perf update. Makes it stable and performant across a wider range of hit rates. Still not quite up to par with round-robin in the general setting, but getting close.

1. Do not normalize waiting requests. 
Hand-wavy intuition is as such: if `waiting_requests` is to be a proxy of the token load for `max_num_batched_tokens` budget, then only the absolute deltas among workers should matter, instead of the relative deltas.

2. Perform predictive updates of `waiting_requests` as done in Rust land. 
A small side-note unrelated to this PR: the  predictive update of `active_kv_blocks` does not seem to be hooked up directly to the Rust routing scheduler, which uses `gpu_cache_usage_perc`

predictive_unnormalized_waiting
(not shown in plot, but normalized + predictive waiting load updates also gave unstable results)

Few more updates:

1. Normalized the overlap score by `kv_total_blocks`, not ISL, to make the units consistent with `gpu_usage_perc`
2. Introduced a softmax sampling on the worker logits to reduce thrashing, which was shown to improve stability and performance empirically

## Summary by CodeRabbit

- **New Features**
  - Improved worker selection by introducing predictive tracking of waiting requests, enhancing request routing responsiveness.

- **Refactor**
  - Updated the handling of waiting requests metrics to use raw values instead of normalized counts, simplifying the routing logic.

### CODE REVIEW

File: examples/llm/components/kv_router.py [hunk 1]
```diff
@@ -220,6 +242,10 @@ def _cost_function(
             for message in log_messages:
                 logger.info(message)
 
+            # Increment predictive waiting for the selected worker before returning
+            predictive_value = self.num_requests_waiting_dict[best_worker_id][1]
```
Comment (jthomson04) [inline examples/llm/components/kv_router.py]: Can't you just do `self.num_requests_waiting_dict[best_worker_id][1] += 1`

File: examples/llm/components/kv_router.py [hunk 2]
```diff
@@ -117,8 +117,29 @@ async def async_init(self):
...
+        # Initialize the predictive waiting requests dictionary
+        # [old_value, predictive_value]
+        self.num_requests_waiting_dict = {}
+        worker_ids = self.workers_client.instance_ids()
+        for worker_id in worker_ids:
+            self.num_requests_waiting_dict[worker_id] = [0.0, 0.0]
+
         logger.info("KV Router initialized")
 
+    def _update_and_get_waiting_value(
```
Comment (jthomson04) [inline examples/llm/components/kv_router.py]: Just to make sure I understand, the logic here is saying: If we have a newly polled value, use that. Otherwise, use the previous value adjusted for any newly routed requests.

**Review by jthomson04 — APPROVED** [review]

File: examples/llm/components/kv_router.py (hunk 1 above)
Comment (PeaBrane) [reply examples/llm/components/kv_router.py]: Ahh yes good catch

File: examples/llm/components/kv_router.py (hunk 2 above)
Comment (PeaBrane) [reply examples/llm/components/kv_router.py]: Yes exactly. I believe that's what the rust kv scheduler is doing, so mirroring the logic. But I should double check. Will add a docstring

**Review by ishandhanani — APPROVED** [review]
